### PR TITLE
HMR guide rewrite

### PR DESCRIPTION
**Pull request by sbaidon**
This PR contains some changes to the `HMR` guide:

-  Update examples to be consistent with `Development` guide

### CODE REVIEW

File: content/guides/hot-module-replacement.md [hunk 1]
```diff
@@ -18,125 +19,194 @@ related:
...
+    filename: '[name].bundle.js',
+    path: path.resolve(__dirname, 'dist')
   }
 };
 ```
 
-Not too bad, huh? Let's test it out using `module.hot.accept`...
+You can also use the CLI to modify the [webpack-dev-server](https://github.com/webpack/webpack-dev-server) configuration with the following command: `webpack-dev-server --hotOnly`.
+
+Now let's update the `index.js` file so that when a change inside `print.js` is detected we tell webpack to accept the updated module.
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: I think you need to mention that the reader should `npm start` their `webpack-dev-server` before altering the `print.js` module.

File: content/guides/hot-module-replacement.md [hunk 2]
```diff
@@ -18,125 +19,194 @@ related:
...
+  btn.innerHTML = 'Click me and check the console!';
+  btn.onclick = printMe;
+
+  element.appendChild(btn);
+
+  return element;
+}
+
+document.body.appendChild(component());
 ```
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: You might be able to just leave out this whole section as nothing has changed from the `development` guide.

File: content/guides/hot-module-replacement.md [hunk 3]
```diff
@@ -18,125 +19,194 @@ related:
...
++ main.js:4330 [HMR] Consider using the NamedModulesPlugin for module names.
 ```
 
-
 ## Gotchas
 
-Hot Module Replacement can be tricky. For example, let's say I have the following class:
+Hot Module Replacement can be tricky. To show this, let's go back to our working example. If you go ahead and click the button on the example page, you will realize the console is printing the old `printMe` function.
+
+This is happening because the button's `onclick` event handler is still bind to the original `printMe` function.
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: /s/still bind/still bound

File: content/guides/hot-module-replacement.md [hunk 4]
```diff
@@ -18,125 +19,194 @@ related:
...
-    console.log('Logging some text: ', text)
-  }
+import _ from 'lodash';
+import printMe from './print.js';
+
+if (module.hot) {
+    module.hot.accept('./print.js', function() {
+    console.log('Accepting the updated printMe module!');
+    document.body.removeChild(element);
+    element = component(); // update binding now that new printMe module has been acepted
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: I would rewrite the comment as:

> Re-render the "component" to update the click handler.

File: content/guides/hot-module-replacement.md [hunk 5]
```diff
@@ -18,125 +19,194 @@ related:
...
 ``` js
-class Logger {
-  log(text) {
-    console.log('Logging some text: ', text)
-  }
+import _ from 'lodash';
+import printMe from './print.js';
+
+if (module.hot) {
+    module.hot.accept('./print.js', function() {
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: This line is indented two tabs instead of one. However, I think this whole block should be changed to a `diff` and indented two extra lines.

File: content/guides/hot-module-replacement.md [hunk 6]
```diff
@@ -18,125 +19,194 @@ related:
...
-  },
-  // ...
+  return element;
 }
+
+document.body.appendChild(element);
+```
+
+This is just one example, but there are many others that can easily trip people up. Luckily, there are a lot of loaders out there, some mentioned below, that will make using this process much easier.
+
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: Add an extra line break between sections (similar to my other `h2` comment above).

File: content/guides/hot-module-replacement.md [hunk 7]
```diff
@@ -18,125 +19,194 @@ related:
...
+## HMR with Stylesheets
+
+Hot Module Replacement with CSS is actually fairly straightforward with the help of the `style-loader`. This loader uses `module.hot.accept` behind the scenes to patch `<style>` tags when CSS dependencies are updated.
+
+Now let's update the configuration file to make use of the loader.
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: You should probably add a quick installation line before the config changes, e.g.

``` bash
npm install --save-dev style-loader css-loader
```

Comment (skipjack) [inline content/guides/hot-module-replacement.md]: Also below the installation line you should put the name of the file that's being edited for consistency with the other guides e.g.

__webpack.config.js__

``` diff
...
```

File: content/guides/hot-module-replacement.md [hunk 8]
```diff
@@ -152,7 +222,6 @@ body {
 
 Change the style on `body` to `background: red;` and you should immediately see the page's background color change without a full refresh.
 
-
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: Same thing re line breaks.

File: content/guides/hot-module-replacement.md (hunk 7 above)
Comment (sbaidon) [inline content/guides/hot-module-replacement.md]: I actually thought about this, but I noticed that on the previous examples `style-loader` and `css-loader` are already added as dev-dependencies, should I write this just as a reminder in case they haven't been installed?

**Comment by sbaidon** [comment]
Thanks for the review will submit changes ASAP!

File: content/guides/hot-module-replacement.md (hunk 7 above)
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: I think it's only mentioned in `asset-management` which is then [cleaned up](https://webpack.js.org/guides/asset-management/#wrapping-up) at the end. I would mention it again just as a reminder.

File: content/guides/hot-module-replacement.md [hunk 9]
```diff
@@ -18,125 +19,194 @@ related:
...
-  },
-
-  devServer: {
-    hot: true, // Tell the dev-server we're using HMR
-    contentBase: path.resolve(__dirname, 'dist'),
-    publicPath: '/'
+    filename: '[name].bundle.js',
+    path: path.resolve(__dirname, 'dist')
   }
 };
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Yeah just use 2 spaces for indentation and 2 spaces at the start of each line in a diff block for the `+` and `-`.

File: content/guides/hot-module-replacement.md [hunk 10]
```diff
@@ -18,125 +19,194 @@ related:
...
++     new webpack.HotModuleReplacementPlugin()
   ],
-
   output: {
-    filename: 'main.js',
-    path: path.resolve(__dirname, 'dist'),
-    publicPath: '/'
-  },
-
-  devServer: {
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Isn't the dev server required for HMR?

File: content/guides/hot-module-replacement.md [hunk 11]
```diff
@@ -18,125 +19,194 @@ related:
...
 
-Not too bad, huh? Let's test it out using `module.hot.accept`...
+You can also use the CLI to modify the [webpack-dev-server](https://github.com/webpack/webpack-dev-server) configuration with the following command: `webpack-dev-server --hotOnly`.
+
+Now let's update the `index.js` file so that when a change inside `print.js` is detected we tell webpack to accept the updated module.
 
 __index.js__
 
 ``` js
-import Library from './library';
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: make this a diff and use the `+`and `-` for consistency :)

File: content/guides/hot-module-replacement.md [hunk 12]
```diff
@@ -18,125 +19,194 @@ related:
     url: /api/hot-module-replacement
 ---
 
+T> This guide extends on code examples found in the [`Development`](/guides/development) guide.
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Can you remove the backticks from "[`Development`]"? We use them for code, but decided to not use them for links in the guides.

File: content/guides/hot-module-replacement.md [hunk 13]
```diff
@@ -18,125 +19,194 @@ related:
...
-    ]
-  },
-  // ...
+  return element;
 }
+
+document.body.appendChild(element);
+```
+
+This is just one example, but there are many others that can easily trip people up. Luckily, there are a lot of loaders out there, some mentioned below, that will make using this process much easier.
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: I would suggest to change this slightly for a nicer sentence:

`This is just one example, but there are many others that can easily trip people up. Luckily, there are a lot of loaders out there (some of which are mentioned below) that will make hot module replacement much easier.`

File: content/guides/hot-module-replacement.md (hunk 7 above)
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: I agree, it's also best to make sure that people who don't follow all the guides in succession are still able to fairly easily follow along with this particular guide.

I'd rather be a bit too verbose in these cases.

File: content/guides/hot-module-replacement.md [hunk 14]
```diff
@@ -18,125 +19,194 @@ related:
...
+      new webpack.HotModuleReplacementPlugin()
+    ],
+    output: {
+      filename: '[name].bundle.js',
+      path: path.resolve(__dirname, 'dist')
+    }
+  };
 ```
 
 hot loading stylesheets is a breeze...
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: I would change this slightly to keep the tone a little more friendly, maybe:

`Hot loading stylesheets is as easy as importing them into a module:`

File: content/guides/hot-module-replacement.md [hunk 15]
```diff
@@ -18,125 +19,194 @@ related:
...
 
 __index.js__
 
 ``` js
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Make this a diff

File: content/guides/hot-module-replacement.md [hunk 16]
```diff
@@ -152,7 +222,6 @@ body {
 
 Change the style on `body` to `background: red;` and you should immediately see the page's background color change without a full refresh.
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Would also include a diff here for the stylesheet.

File: content/guides/hot-module-replacement.md (hunk 15 above)
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: Can you also add in a diff for the project structure (you're adding a new file, right?), and put the new file contents in a code block as well. Unless the file is already present, then a diff is sufficient (but I think it's not there anymore, as one of the guides does a cleanup at the end).

File: content/guides/hot-module-replacement.md (hunk 2 above)
Comment (sbaidon) [inline content/guides/hot-module-replacement.md]: I did add the `module.hot` part, or do you mean doing something like this:

``` js
import _ from 'lodash';
import printMe from './print.js';

if (module.hot) {
    module.hot.accept('./print.js', function() {
    console.log('Accepting the updated printMe module!');
    printMe();
  })
}
...
```

Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: No what you did was fine IMO, just make it a full diff ;)

**Comment by sbaidon** [comment]
Just pushed the changes

Hope I didn't miss anything!

**Comment by evolve2k** [comment]
I today opened a PR to request ehnancements to this documentation, @TheDutchCoder just kindly redirected me here. Glad to see this rewrite in progress. 
I've had a quick look over the commit history but it's hard to tell at a glance if my thoughts have been addressed.. So here's what I said earlier. If they are already addressed that's great! If not please feel free to account for them.

## Feedback on current live documentation

Progressing through the tutorial so far, each tute lead on nicely to the next.
This one however breaks the flow and provided no clues as to which lines to add like previous files do.

Basically the convention of the green lines beginning with + signs is not present in this file, breaking the flow of the tutorial at this point.

**Comment by TheDutchCoder** [comment]
@evolve2k Yeah we're doing a major guides overhaul. Some of them are "done" and in place, but a couple of them are still WIPs and don't follow the same "flow" as the ones that are done.

It's our goal to make each guide smoothly follow up on a previous one.

If you find any of them lacking, please let us know (and even open a PR if you like), we're always happy with contributions!

File: content/guides/hot-module-replacement.md [hunk 17]
```diff
@@ -195,33 +209,82 @@ Now let's update the configuration file to make use of the loader.
...
+    | - print.js
++   | - styles.css
 ```
 
 __styles.css__
 
 ``` css
 body {
-  background: blue;
+  background: red;
```
Comment (TheDutchCoder) [inline content/guides/hot-module-replacement.md]: This should be `background: blue;` to make it line up with the diff further down ;)

**Comment by skipjack** [comment]
@evolve2k yeah you can check out [this comment](https://github.com/webpack/webpack.js.org/issues/1258#issuecomment-312139191) on #1258 for more information on the current guides work. I've been keeping it updated based on the work and PRs submitted.

File: content/guides/hot-module-replacement.md [hunk 18]
```diff
@@ -18,140 +19,271 @@ related:
...
+You can also use the CLI to modify the [webpack-dev-server](https://github.com/webpack/webpack-dev-server) configuration with the following command: `webpack-dev-server --hotOnly`.
 
-  devServer: {
-    hot: true, // Tell the dev-server we're using HMR
-    contentBase: path.resolve(__dirname, 'dist'),
-    publicPath: '/'
-  }
-};
-```
+To get it up and running let's do `npm start` from the command line.
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: I would rephrase "let's do..." => "let's run..."

File: content/guides/hot-module-replacement.md [hunk 19]
```diff
@@ -18,140 +19,271 @@ related:
...
-Hot Module Replacement can be tricky. For example, let's say I have the following class:
+Hot Module Replacement can be tricky. To show this, let's go back to our working example. If you go ahead and click the button on the example page, you will realize the console is printing the old `printMe` function.
+
+This is happening because the button's `onclick` event handler is still bound to the original `printMe` function.
 
-``` js
-class Logger {
-  log(text) {
-    console.log('Logging some text: ', text)
+To make this work with HMR we need to update that binding to the new `printMe` function using `module.hot.accept`
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: You should probably to add some punctuation to the end of this sentence (e.g. `:`).

File: content/guides/hot-module-replacement.md [hunk 20]
```diff
@@ -18,140 +19,271 @@ related:
...
-__index.js__
 
-``` js
-import Lib from './library';
-import './styles.css';
+## HMR with Stylesheets
+
+Hot Module Replacement with CSS is actually fairly straightforward with the help of the `style-loader`. This loader uses `module.hot.accept` behind the scenes to patch `<style>` tags when CSS dependencies are updated.
+
+First let's install both loaders with the following command: `npm install --save-dev style-loader css-loader`
```
Comment (skipjack) [inline content/guides/hot-module-replacement.md]: You should break the command to a new line e.g.

```` md
First let's install both loaders with the following command:

``` bash
npm install --save-dev style-loader css-loader
```
````

**Comment by sbaidon** [comment]
Just made the changes, hope everything is alright now 👍

**Comment by skipjack** [comment]
@TheDutchCoder I'll leave this to you to "Squash and Merge" once the build checks finish.

**Comment by TheDutchCoder** [comment]
Yes sir, will do!

**Comment by TheDutchCoder** [comment]
@skipjack was that a better  "Squash & Merge"? :)

**Comment by sbaidon** [comment]
I just noticed I was missing a lodash import in a diff block, I opened up a new PR #1456  is that ok?

**Comment by skipjack** [comment]
@TheDutchCoder beautiful 🚀  🎉  😆 ... feel free to also drop the bullet points from the body, e.g.

``` md
docs(guides): rewrite of the HMR guide (#1439)

Rewrite of the hot-module-replacement guide.
```

@sbaidon that's perfect, thanks for following up!

**Comment by TheDutchCoder** [comment]
Yeah no problem!

On Jul 26, 2017 12:07 PM, "Sergio Emir Baidon Carrillo" <
notifications@github.com> wrote:

I just noticed I was missing a lodash import in a diff block, I opened up a
new PR #1456 <https://github.com/webpack/webpack.js.org/pull/1456> is that
ok?

—
You are receiving this because you modified the open/close state.

Reply to this email directly, view it on GitHub
<https://github.com/webpack/webpack.js.org/pull/1439#issuecomment-318101862>,
or mute the thread
